### PR TITLE
Globally Unique FATE Transaction Ids - Part 3

### PR DESCRIPTION
**Pull request by kevinrr888**
This addresses several previously deferred changes for issue #4044. Root of most of these new changes were from changing TabletMetadata and TabletUpdates (in Ample) - FateId is now stored in the Metadata table instead of just the formatted long id. Addresses deferred changes to TabletMetadata, TabletUpdates, CompactionConfigStorage, SelectedFiles, and Ample.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java:101]: We may need a follow on update for upgrade

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java:115]: The following would probably make mch more readable json and be tolerant of changes to the implementation of FateId like variable renames.

```suggestion
    jData.fateId = fateId.canonical();
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/CompactionMetadata.java:125]: This suggestion goes with another comment.

```suggestion
        CompactorGroupIdImpl.groupId(jData.groupId), jData.propDels, FateId.from(jData.fateId));
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java:342]:
```suggestion
        // ELASTICITY_TODO issue 4044 will this be an issue removing the old code?  May need to introduce code in upgrade to handle old format.
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/SelectedFiles.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/SelectedFiles.java:65]:
```suggestion
      out.name("fateId").value(selectedFiles.getFateId().canonical());
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/SelectedFiles.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/SelectedFiles.java:92]:
```suggestion
          case "fateId":
```

File: server/base/src/main/java/org/apache/accumulo/server/compaction/CompactionConfigStorage.java
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/compaction/CompactionConfigStorage.java:107]:
```suggestion
      String[] fields = child.split(DELIMITER);
      Preconditions.checkState(fields.length == 2, "Unexpected child %s", child);
```

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementParameters.java
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementParameters.java:209]: Making the key type for this map a string that is the canonical value would yield nicer and more stable json.

File: test/src/main/java/org/apache/accumulo/test/functional/AmpleConditionalWriterIT.java
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/AmpleConditionalWriterIT.java:759]: Reordered the vars to be sorted.

```suggestion
      FateId fateId45L = FateId.from(type, 45L);
      FateId fateId55L = FateId.from(type, 55L);
      FateId fateId56L = FateId.from(type, 56L);
      FateId fateId65L = FateId.from(type, 65L);
      FateId fateId75L = FateId.from(type, 75L);
```

File: test/src/main/java/org/apache/accumulo/test/functional/SplitRecoveryIT.java
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/SplitRecoveryIT.java:50]: #4260 will conflict with these changes,  will hold off on merging it until this PR is merged.

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java:115]: Hard to tell from looking at this diff, but I assume the types line up?  Meaning `entry.getValue()` is a FateId type.

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementParameters.java
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementParameters.java:209]: If this seems like a big change, could add an ELASTICITY_TODO for this to get this PR merged.

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java
Comment (kevinrr888) [reply server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java:115]: Yes, types line up. TabletMetadata.getLoaded() (tablet.getLoaded()) returns Map<StoredTabletFile, FateId>

Comment (keith-turner) [reply server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/CleanUpBulkImport.java:115]: Thanks

**Comment by kevinrr888** [comment]
Completed review changes:
- FateId refactored (better Pattern for FateId and simpler validation)
- CompactionMetadata GSonData changed to use the canonical FateId string instead of FateId (note that old data may need to be considered here)
- SelectedFiles renamed "txid" --> "fateId". SelectedFilesTest updated to be consistent with changes
- CompactionConfigStorage added Preconditions check
- Changed compactionHints in TabletManagementParameters from Map<FateId,Map<String,String>> --> Map<String,Map<String,String>>: required changes to TabletManagementParameters, CompactionConfigStorage, CompactionJobGenerator, TabletManagementParametersTest, Manager, CancelCompactions, and PreDeleteTable
- AmpleConditionalWriterIT.testCompacted() reordering of FateIds

**Comment by keith-turner** [comment]
> Changed compactionHints in TabletManagementParameters from Map<FateId,Map<String,String>> --> Map<String,Map<String,String>>: required changes to TabletManagementParameters, CompactionConfigStorage, CompactionJobGenerator, TabletManagementParametersTest, Manager, CancelCompactions, and PreDeleteTable

Sorry I did not communicate very clearly for the comment about that.  Was not looking to change the external map type parameters for TabletManagementParameters to have a key type of String.  Just thought it would be good to change internal map used for json so that the json encoding would cleaner.  So was thinking  of pattern like the following.

 * The public methods on TabletManagementParameters use `Map<FateId,Map<String,String>>`
 * Internally for json encoding `Map<FateId,Map<String,String>>` is converted to/from `Map<String,Map<String,String>>`

**Comment by kevinrr888** [comment]
> > Changed compactionHints in TabletManagementParameters from Map<FateId,Map<String,String>> --> Map<String,Map<String,String>>: required changes to TabletManagementParameters, CompactionConfigStorage, CompactionJobGenerator, TabletManagementParametersTest, Manager, CancelCompactions, and PreDeleteTable
> 
> Sorry I did not communicate very clearly for the comment about that. Was not looking to change the external map type parameters for TabletManagementParameters to have a key type of String. Just thought it would be good to change internal map used for json so that the json encoding would cleaner. So was thinking of pattern like the following.
> 
> * The public methods on TabletManagementParameters use `Map<FateId,Map<String,String>>`
> * Internally for json encoding `Map<FateId,Map<String,String>>` is converted to/from `Map<String,Map<String,String>>`

Ah gotcha. Will fix.

**Review by keith-turner — APPROVED** [review]

**Comment by dlmarion** [comment]
@kevinrr888 - The latest IT run is failing (stacktrace below). I think it might be due to this merge. Just an FYI...

```
org.opentest4j.AssertionFailedError: Test json should be identical to actual metadata at this point ==> expected: <{"fateId":"FATE:USER:0000000000000002","selAll":true,"files":["{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000070.rf\",\"startRow\":\"\",\"endRow\":\"\"}","{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000071.rf\",\"startRow\":\"\",\"endRow\":\"\"}","{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000072.rf\",\"startRow\":\"\",\"endRow\":\"\"}"]}> but was: <{"txid":"FATE:USER:0000000000000002","selAll":true,"files":["{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000070.rf\",\"startRow\":\"\",\"endRow\":\"\"}","{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000071.rf\",\"startRow\":\"\",\"endRow\":\"\"}","{\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000072.rf\",\"startRow\":\"\",\"endRow\":\"\"}"]}>
```

**Comment by kevinrr888** [comment]
> @kevinrr888 - The latest IT run is failing (stacktrace below). I think it might be due to this merge. Just an FYI...
> 
> ```
> org.opentest4j.AssertionFailedError: Test json should be identical to actual metadata at this point ==> expected: 
> <{"fateId":"FATE:USER:0000000000000002","selAll":true,"files":["
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000070.rf\",\"startRow\":\"\",\"endRow\":\"\"}","
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000071.rf\",\"startRow\":\"\",\"endRow\":\"\"}","
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000072.rf\",\"startRow\":\"\",\"endRow\":\"\"}"]}> 
> but was: <{"txid":"FATE:USER:0000000000000002","selAll":true,"files":["
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000070.rf\",\"startRow\":\"\",\"endRow\":\"\"}","
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000071.rf\",\"startRow\":\"\",\"endRow\":\"\"}","
> {\"path\":\"hdfs://localhost:8020/accumulo/tables/2a/default_tablet/F0000072.rf\",\"startRow\":\"\",\"endRow\":\"\"}"]}>
> ```

Thanks for pointing this out. I created https://github.com/apache/accumulo/pull/4265

**Pull request merged**
